src/simplegmres.jl

```diff
@@ -0,0 +1,600 @@
+
+    T = eltype(x)
+    q = PlisI ? w : q
+    r₀ = PlisI ? w : q
+    xr = restart ? Δx : x
+
+    if β == 0
+        return SciMLBase.build_linear_solution(lincache.alg, x, r₀, lincache;
+            retcode = ReturnCode.Success)
+    end
+
+    rNorm = β
+    npass = 0        # Number of pass
+
+    iter = 0        # Cumulative number of iterations
+    inner_iter = 0  # Number of iterations in a pass
+
+    # Tolerance for breakdown detection.
+    btol = eps(T)^(3 / 4)
+
+    # Stopping criterion
+    breakdown = false
+    inconsistent = false
+    solved = rNorm ≤ ε
+    inner_maxiters = maxiters
+    tired = iter ≥ maxiters
+    inner_tired = inner_iter ≥ inner_maxiters
+    status = ReturnCode.Default
+
+    while !(solved || tired || breakdown)
+        # Initialize workspace.
+        nr = 0  # Number of coefficients stored in Rₖ.
+
+        if restart
+            xr .= zero(T)  # xr === Δx when restart is set to true
+            if npass ≥ 1
+                mul!(w, A, x)
+                axpby!(one(T), b, -one(T), w)
+                PlisI || ldiv!(r₀, Pl, w)
+            end
+        end
+
+        # Initial ζ₁ and V₁
+        β = _norm2(r₀)
+        z[1] = β
+        V[1] .= r₀ / β
+
+        npass = npass + 1
+        inner_iter = 0
+        inner_tired = false
+
+        while !(solved || inner_tired || breakdown)
+            # Update iteration index
+            inner_iter += 1
+            # Update workspace if more storage is required and restart is set to false
+            if !restart && (inner_iter > memory)
+                append!(R, zeros(T, inner_iter))
+                push!(s, zero(T))
+                push!(c, zero(T))
+            end
+
+            # Continue the Arnoldi process.
+            p = PrisI ? V[inner_iter] : p
+            PrisI || ldiv!(p, Pr, V[inner_iter])  # p ← Nvₖ
+            mul!(w, A, p)                         # w ← ANvₖ
+            PlisI || ldiv!(q, Pl, w)                 # q ← MANvₖ
+            for i in 1:inner_iter
+                R[nr + i] = dot(V[i], q)       # hᵢₖ = (vᵢ)ᴴq
+                axpy!(-R[nr + i], V[i], q)     # q ← q - hᵢₖvᵢ
+            end
+
+            # Compute hₖ₊₁.ₖ
+            Hbis = _norm2(q)  # hₖ₊₁.ₖ = ‖vₖ₊₁‖₂
+
+            # Update the QR factorization of Hₖ₊₁.ₖ.
+            # Apply previous Givens reflections Ωᵢ.
+            # [cᵢ  sᵢ] [ r̄ᵢ.ₖ ] = [ rᵢ.ₖ ]
+            # [s̄ᵢ -cᵢ] [rᵢ₊₁.ₖ]   [r̄ᵢ₊₁.ₖ]
+            for i in 1:(inner_iter - 1)
+                Rtmp = c[i] * R[nr + i] + s[i] * R[nr + i + 1]
+                R[nr + i + 1] = conj(s[i]) * R[nr + i] - c[i] * R[nr + i + 1]
+                R[nr + i] = Rtmp
+            end
+
+            # Compute and apply current Givens reflection Ωₖ.
+            # [cₖ  sₖ] [ r̄ₖ.ₖ ] = [rₖ.ₖ]
+            # [s̄ₖ -cₖ] [hₖ₊₁.ₖ]   [ 0  ]
+            (c[inner_iter], s[inner_iter], R[nr + inner_iter]) = _sym_givens(R[nr + inner_iter],
+                Hbis)
```

Comment:
Aren't there Base functions for this?

Reply:
Seems like there is. I have opened an issue to track those https://github.com/SciML/LinearSolve.jl/issues/375

Reply:
@avik-pal This a 2x2 Givens reflection and not a 2x2 Givens rotation.
This a not the same thing.
The advantage of the reflections is that they are Hermitian and it's easier to perform products with `Q'`.